### ISSUE TITLE
Add support for Microsoft Fabric

### ISSUE THREAD

**Issue by mattiasthalen** [TobikoData/sqlmesh#3374]
Microsoft Fabric is becoming more and more popular with our clients. And most of the time it's new projects or migrations. I.e., the perfect time to implement SQLMesh.

I know that there are some issues with their lakehouse entity and creating schemas / tables. And that's why dbt only work on their warehouse entity.

Anyway, I hope this is something you'd be willing to look into!

**Comment by tobymao** [comment]
@treysp which limitations do you think are critical? it seems like if the user doesn't do certain operations, we won't hit these limitations?

**Comment by treysp** [comment]
The limits on `ALTER TABLE ADD/ALTER/DROP COLUMN` were the main ones that concerned me. 

But you're right that those limitations would be feature-specific (primarily on forward-only), so there's no inherent problem if we just tell people those aren't supported.

Sorry for the premature closing @mattiasthalen!

**Comment by mattiasthalen** [comment]
@treysp No worries!

I'm curious though, are those there for the warehouse object as well? I was under the impression that it only was on the lakehouse.

**Comment by treysp** [comment]
Totally possible I'm misinterpreting that page - this text made me think it was platform-wide:

```
Applies to: ✅ SQL analytics endpoint and Warehouse in Microsoft Fabric

This article covers the T-SQL language syntax capabilities of Microsoft Fabric, when querying the SQL analytics endpoint or Warehouse.
```

**Comment by mattiasthalen** [comment]
@treysp you are absolutely right, I missed that part!

It sure does have it quirks, e.g., supporting UUID as a data type in the lakehouse/warehouse, but not in PowerBI. So much for consistency 😅

**Comment by pycal** [comment]
+1 

our company has to use fabric, and if a connector for fabric is created we will use sqlmesh.

**Comment by tobymao** [comment]
we're planning to work on this soon after the holidays

**Comment by erindru** [comment]
Update on this:

SQLMesh needs some fundamental constructs in order to work properly, namely - schemas, tables and views.

It turns out the Lakehouse entity in Fabric does not support schemas by default. However, if we enable the [preview feature for schemas](https://learn.microsoft.com/en-us/fabric/data-engineering/lakehouse-schemas#create-a-lakehouse-schema) - it stops supporting _views_. This is documented as a [limitation](https://learn.microsoft.com/en-us/fabric/data-engineering/lakehouse-schemas#public-preview-limitations) in the public preview of schemas.

We need schemas for the physical layer and views for the virtual layer, so SQLMesh cannot work with Lakehouse until Microsoft resolves this. According to their documentation:

> They'll be resolved in the coming releases before General Availability.

So we can revisit this after schema-enabled Lakehouses becomes generally available and theoretically support views.

Of course, the Warehouse endpoints that are TSQL-only are unaffected.

I'm curious - is there any demand for Warehouse / TSQL or are people mostly wanting to use Lakehouse / Spark ?

**Comment by cseHdz** [comment]
+1 - agree with @mattiasthalen above

We also would like to use SQLMesh with Warehouse. Lakehouse is a nice to have.
Our workflow is also using Lakehouse as a landing zone, and Warehouse for transforms.

**Comment by mattiasthalen** [comment]
And for reference, you can target the lakehouse via the warehouse endpoint, it's "just" a different database.

E.g.:
lakehouse__landing_zone.dbo.raw__table
warehouse.silver.stg__table

**Comment by odikia** [comment]
+1 also agree with @mattiasthalen above.

we’ll be transitioning to fabric architecture for primary data products in the coming year, and would like to take advantage of sqlmesh across our multi-cloud enterprise.

**Comment by mhofmeester** [comment]
+1 with @mattiasthalen here!

For our clients we offer either a Databricks or Fabric centric approach often with DBT. It would be great if SQLMesh would support the warehouse endpoint, as you advertise to solve some of the limitations/annoyances we faced with DBT.

Fabric (Microsoft/Azure) has quite a dominant presence here in the Netherlands for data engineering capabilities. I figure it would be an interesting userbase. 

With regards to the lakehouse vs. warehouse discussion, ideally SQLMesh would support both eventually (when technically feasible). However, the warehouse endpoint is currently by far the most resource efficient transform solution within Fabric, and would always have my preference for cost reasons.

**Comment by toninator** [comment]
+1

**Comment by pycal** [comment]
our implementation is, lakehouse for bronze, warehouse for silver and gold

we could definitely live with the limitation

**Comment by mattiasthalen** [comment]
Any news yet? ☺️

**Comment by tobymao** [comment]
@mattiasthalen this is not a priority for us at the moment. however, if you're interested, we do offer feature prioritization as a part of our enterprise / paid offering.

**Comment by mattiasthalen** [comment]
We can now use sqlmesh with the sql db object: https://github.com/TobikoData/sqlmesh/pull/4686

I've tried it with a warehouse object, but since it's not pure tsql, there are still some stuff to take care of. Now it's complaining about merge.

I'll probably have a look soonish.

**Comment by fresioAS** [comment]
Linked to PR in #4751

**Comment by travispaice** [comment]
We'd also consider Fabric for future implementations, though the sqlmesh component is more important in a warehouse/sql database context more than the lakehouse.

That said there are issues with sqlmesh compatibility with MS SQL, and Azure SQL, and therefore presumably Fabric DWH + Fabric SQL DB. Specifically relating to macro syntax compatibility (all) and multi-part object naming (Azure SQL).

**Comment by mattiasthalen** [comment]
> We'd also consider Fabric for future implementations, though the sqlmesh component is more important in a warehouse/sql database context more than the lakehouse.
> 
> That said there are issues with sqlmesh compatibility with MS SQL, and Azure SQL, and therefore presumably Fabric DWH + Fabric SQL DB. Specifically relating to macro syntax compatibility (all) and multi-part object naming (Azure SQL).

Me and @fresioAS are working on an engine adapter specifically using he sql endpoint for the warehouse and have come quite far, just some integration tests that needs to pass. And those are mostly because we added some new config inputs (workspace id IIRC) and that's not setup in the CI env. But I think @erindru was to take a look next week or the week after.

I'm also vibe coding a spark adapter for the Lakehouse using the Livy API.

I'm curious, what are the macro issues? Could you perhaps log an issue and tag me in it?

The multipart thing is wonky in fabric. Db switching doesn't really work, MS recommends three part identifiers. bBut drop schema and create view doesn't support multi part identifiers. 🙈

**Comment by travispaice** [comment]
Hey @mattiasthalen , I've linked you into https://github.com/TobikoData/sqlmesh/issues/5034 for more context on the macro syntax issue.

We're not in a rush for Fabric in production due to the numerous bugs and limitations we have encountered in PoC's, but believe these will be resolved in 6-12 months.

Awesome you are working towards it though, are you considering Fabric SQL DB as well or just Fabric DWH? In theory it should be the same end point, but the limitations/features of SQL DB may be different than DWH.
